doc/src/guides/contributing/deprecations.md

```diff
@@ -0,0 +1,387 @@
+require deprecations to change (again, this list is only a general set of
+guidelines).
+
+```{admonition} Not Public API
+- The precise form of an expression. In general, functions may be changed to
+  return a different but mathematically equivalent form of the same
+  expression. This includes a function returning a value which it was not able
+  to compute previously.
+- Positional argument names.
+- Functions and methods that are private, i.e., for internal use only. Such
+  things should generally be prefixed with an underscore `_`, although this
+  convention is not currently universally adhered to in the SymPy codebase.
+- Anything explicitly marked as "experimental".
+- Changes to behavior that were mathematically incorrect previously (in
+  general, bug fixes are not considered breaking changes, because despite the
+  saying, bugs in SymPy are not features).
+- Anything that was added before the most recent release. Code that has not
+  yet made it into a release does not need to be deprecated. If you are going
+  to change the API of new code, it is best to do it before a release is made
+  so that no deprecations are necessary for future releases.
+```
+
+Note: both public and private API functions are included in the [reference
+documentation](reference), and many functions are not included there which
+should be, so this should not be used to determine whether something public or
+not. Cleaning up the distinction between public and private APIs, as well as
+the categorization in the reference documentation is currently an open issue
+for SymPy.
+
+If you're unsure, there is no harm in deprecating something even if it might
+not actually be "public API".
+
+## The purpose of deprecation
+
+Deprecation has several purposes:
+
+- To allow existing code to continue to work for a while, giving people a
+  chance to upgrade SymPy without fixing all deprecation issues immediately.
+- To warn users that their code will break in a future version.
+- To inform users how to fix their code so that it will continue work in
+  future versions.
+
+To this end, a deprecation warning should be something that users can remove
+by fixing their code. Deprecation warnings that fire unconditionally, or which
+can otherwise not effectively be silenced by users other than by using a
+warnings filter should be avoided.
+
+Furthermore, deprecation warnings should inform users of a way to change their
+code so that it works in the same version of SymPy, as well as all future
+versions. Features that do not have a suitable replacement should not be
+deprecated, as there will be no way for users to fix their code to continue to
+work.
+
+Deprecations should always
+
+1. Allow users to continue to use the existing APIs unchanged during the
+   deprecation period (with a warning).
+2. Allow users to always fix their code so that it stops giving the warning.
+3. After users fix their code, it should continue to work after the deprecated
+   code is removed.
```

Comment:
Another point to consider here is how a library that attempts to support multiple versions of sympy should straddle the versions without seeing any warning. If there is a way to do it like:
```python
try:
    from sympy.x import y
except ImportError:
    from sympy.z import y
```
then that should be documented.

Reply:
That's a good point. I think generally users should do version checks. After all, each deprecation lists the first version it is deprecated in. Otherwise they have to do some sort of warning filter, or trying the new way first and falling back to the old way if it doesn't work, which may not be simple in general.

Reply:
The point is that whatever downstream libraries need to do should be clearly documented. Filtering warnings doesn't seem acceptable to me but providing a clear sample of code that can be copied into somewhere does.